---
title: "Part1"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Practical Genome Scan using SNPrelate

In this tutorial we will use a sliding window approach to compute expected heterozygosity and FST along the genome.

```{r}
library(SNPRelate)
```


```{r}


# read the genotype matrix
geno <- as.matrix(read.table("TestGenomeScan.geno", header=T, stringsAsFactors = F, na.strings = "NA"))
# check if we read genotype matrix correctly
str(geno) # matrix with nsites rows and nind columns
# How many individuals? How many sites?

# get the label of the individuals
inds <- colnames(geno)

# vector with the population name
pop.names <- c("Pop1", "Pop2")

# A way to get this for all populations
# go through each pop.name and get the index of individuals of each pop 
sample_inds <- list() # list to save the index of individuals belonging to each pop
for(i in 1:length(pop.names)) {
  # apply the function to detect which individuals belong to each pop
  sample_inds[[i]] <- which(substr(inds, 1, nchar(pop.names[i]))==pop.names[i])
}

npops <- length(pop.names)

# create tag 
# this string will be added to the files saved by SNPrelate
filename_tag <- "GenomeScan"

# create a vector with the population corresponding to each individual
popinds <- names(geno) # initiliaze this with the individual names
for(i in 1:length(pop.names)) {
  # get pop.name label for individuals belonging to a given pop
  popinds[sample_inds[[i]]] <- pop.names[i]
}

# SNP relate - WINDOW SCAN
# create input geno file for package
filename_geno <- paste(filename_tag,"_geno.gds",sep="")
# use the package function, giving as input:
# - genmatrix is the genotype matrix "geno"
# - sample.id is the name of individuals in variable "inds"
snpgdsCreateGeno(gds.fn=filename_geno, genmat=geno, snpfirstdim=TRUE, sample.id=inds)

# Open the GDS file that you just created
# this GDS file is the format used by SNPrelate
genofile <- snpgdsOpen(filename=filename_geno)

# SCAN THE GENOME FOR FST AND EXPECTED HETEROZYGOSITY
# Compute the pairwise FST in a sliding window analysis using function snpgdsFst
windowSize <- 100
WindowShift <- 50 # smaller than window to overlap
fst_sw <- list() # list to save the sliding window FST values for each pair of populations
allfreq_sw <- list() # list to save the sliding window allele frequency values
# go through each pair of populations
for(i in 1:(length(pop.names))) {
  fst_sw[[i]] <- list()
  if(i < length(pop.names)) {
    for(j in (i+1):length(pop.names)) {
      # select individuals from the two populations
      keepfst <- popinds==pop.names[i] | popinds==pop.names[j] 
      # Perform a sliding windon analysis of FST
      fst_sw[[i]][[j]] <- snpgdsSlidingWindow(genofile, winsize=windowSize, shift=WindowShift, FUN="snpgdsFst", as.is="numeric", sample.id=inds[keepfst], population=as.factor(popinds[keepfst]))
    }  
  }
  # perform the sliding window analysis to get the allele frequencies
  allfreq_sw[[i]] <- snpgdsSlidingWindow(genofile, winsize=windowSize, shift=WindowShift, snp.id=NULL, remove.monosnp=FALSE,
                                    FUN="snpgdsSNPRateFreq", as.is="numeric", sample.id = inds[sample_inds[[i]]])
  
}

# Plot the heterozygosity along the chromosome
# from the allele frequencies obtain expected heterozygosity
exp_het <- matrix(NA, ncol=length(allfreq_sw[[1]]$chr1.val), nrow=npops)
for(i in 1:npops) {
  exp_het[i,] <- (2*allfreq_sw[[i]]$chr1.val)-(2*allfreq_sw[[i]]$chr1.val^2)
}  
# plot the expected heterozygosity for each population  
plot(exp_het[1,], type="l", ylim=c(0,0.4), ylab="Expected heterozygosity", xlab="SNP position (index)")
for(i in 2:(length(pop.names))) {
  lines(exp_het[i,], type="l", col=2)
}

# Plot the pairwise FST value
par(mfrow=c(1,1))
plot(fst_sw[[1]][[2]]$chr1.val, type="l", ylim=c(-0.1,0.6), xlab="SNP position", ylab="FST")

# Close the GDS file
snpgdsClose(genofile)
```



Exercise

Modify the code above to test different window sizes (windowSize and WindowShift variables).

TO DISCUSS:

- Is there evidence of variation in the genetic diversity along the genome?

* There's quite some variation along the genome (black and red graph) 


- Is there evidence of variation in the genetic differentiation along the genome?

* Yes, we can see that there's some peaks of high Fst 


- What is the effect of changing the window size and window shift?

* We have more details. Finer variations but also much more noisy. 


- Based on these results do you think there is evidence for a region under selection?

We have differenciation and also genetic diversity that can be good indication for regions under selection. Pop1 in black and 2 in red: we can see at the right extremity, we can see a reduced diversity in pop1. 

In differenciation (second graph) we can see a peak at start and end. The strongest indication for a region under selection would be at the end of the chromosome. The selection is applied on pop1 (black) because the diversity decrease on a specific area. 


- How can you be sure it is due to selection and just not the random stochasticity of genetic drift?


Compare the region that has signature of selection with the rest of the genome. The peak on the right is much higher compared to the rest of the genome -> not just drift. 













































































