---
title: "ExerciseEvogeno"
author: "Thibault Schowing"
date: "12/03/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



```{r cars}
source("coalfunctions.r")
```

## Exercise 1



```{r pressure}
# load functions defined in "coalfunctions.r" that will be used to simulate and plot coalescent trees
source("coalfunctions.r")

# call sim.tree() function to simulate gene trees

## In this example, the population size is 1000
# INPUT OF sim.tree() function
# sample - how many gene copies (lineages) are sampled?
# current - what is the current effective size (in number of haploid gene copies)?
# ancestral - what is the ancestral effective size (in number of haploid gene copies)?
# time - how long ago did the size changed from ancestral to current?
# nrep - how many loci?  
nloci <- 10 # number of loci
nind <- 2 # number of individuals (multiplied by 2 after for diploÃ¯ds)
# simulate a tree with function sim.tree()
tree <- sim.tree(sample = 2*nind, current = 1000, ancestral=1000, time=0, nrep=nloci)

## The function draw.tree() plots (on the screen) the coalescent tree
## the nrep is the number of trees simulated.
draw.tree(tree, nrep=nloci)

## You can get the coalescent time intervals with
## the function coaltimes()
## the nrep is the number of trees simulated.
coal <- coaltimes(tree, nrep=nloci)

# the output is a matrix where row is the coalescent interval
# and each column is a tree. The 1st column has the number of lineages in each interval.

# To get the time in generations, you need to multiply by 4Ne
coal[,-1]*4*1000 
```



## Questions

* Simulate 10 trees (change nloci) of a sample of 2 diploid individuals from a population with size 1000 and plot them. How similar are the 10 trees?

----------------------------------

The 10 trees are very different even though the first coalescence happen very early and take longer to happen as far in time we go. The variance is big as expected. 

----------------------------------

* These 10 trees represent the gene genealogy at 10 independent sites. Given the gene trees do you expect similar genetic diversity at the 10 sites of the genome?


-----

No, with random mating there is a huge variation in the gene trees _(Figure 4.2, Hein et al (2004) Gene Genealogies, Variation and Evolution.)_

-----


* Do the time to MRCA and total branch length follow the theoretical expectations?


-----

Just by looking at the graphs, we can see that indeed the coalescence time is very short in the beginning and gets longer and longer the further back in time we go. 

-----



# Exercise 2

```{r}
# Add mutations to the trees with constant, population decline and population expansion
# mu is the mutation rate per site
# L  is the number of sites per locus
nloci <- 10 # number of loci
nind <- 5 # number of diploid individuals
# simulate trees and mutations with function sim.tee.mut()
# INPUT: 
#    mu: mutation rate per site per generation
#    L:  total number of sites




```


### Constant population

```{r}
# constant popultaion
muttree <- sim.tree.mut(sample=2*nind, current=1000, ancestral = 1000, time = 0, nrep = nloci, mu=1e-7, L=1e5)
# plot the tree and haplotypes
plot.haplotypes(muttree, nrep=nloci)
```




### Bottleneck - population decline
```{r}
# bottleneck (population decline)
muttree_bot <- sim.tree.mut(sample=2*nind, current=100, ancestral = 1000, time = 50, nrep = nloci, mu=1e-7, L=1e5)
# plot the tree and haplotypes
plot.haplotypes(muttree_bot, nrep=nloci)


```



### Expanding population

```{r}
# population expansion
muttree_exp <- sim.tree.mut(sample=2*nind, current=10000, ancestral = 100, time = 1000, nrep = nloci, mu=1e-7, L=1e5)
# plot the tree and haplotypes
plot.haplotypes(muttree_exp, nrep=nloci)

```

----
How much variation is there among trees from different loci? Do we expect similar patterns across different sites of the genome?
----

They are very different ! -> genetic drift, mutation and random topology. Each region of the genome have a different gene tree.

----
What is the effect of a bottleneck in the gene trees and sharing of mutations across individuals?
----

You dont' expect many shared mutations and rare singletons because the individual lineage branches are very short.
Also some variation. Typical: lots of recent coalecence events -> will be very similar (share mutations).

----
What is the effect of a population expansion in the gene trees and sharing of mutations across individuals?
----

We see exactely the opposite, the population is so large today, that it is very likely that they have a far away MRCA. Most mutation are unique (singletons). Very long external branches. 

----
What is the scenario leading to more mutations (SNPs)?
----

Expanding population have more and longer branches and this leads to more mutations (unique). Constant population size can have more mutation because the process is stochastic.



----
What is the scenario leading to higher genetic diversity?
----

Same, the longer the branches, the more mutation we expect. We expect less diversity in bottleneck populations because mutations are shared (recent coalescence events, mutation on long past branches).

Answer: depends how we define genetic diversity (same as above in this case if it's just the nb of mutation). We can have other measurment for genetic diversity.  



----
Can we use real data to distinguish these scenarios?
----

Yes if we have enough information: genetics, history of the population (e.g. migration that lead to a bottleneck etc).

Answer: we don't necessarly need time series data. We can juste use today's sequences. We can have prediction to observe what we expect to observe. Gives a prediction for an observed sequence data (what?!)



