---
title: "Evo_genomics_Genetic drift"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Practical session: forward simulations under the Wright-Fisher model

Vitor C. Sousa

March 2020

## Genetic drift, effective size and selection
### Objectives of this practical session
Simulate trajectories of allele frequency change through time under a Wright-Fisher model.

Characterize the effect of demographic factors (effective population size and bottleneck) in the allele frequency change, and probability of loss and fixation of a neutral allele.

Characterize the effect of directional selection on the allele frequency change, and probability of loss and fixation of a beneficial allele.

Examine the fate of a beneficial allele under positive (directional selection) with a given selective coefficient s in populations of different effective sizes.

GOAL: Simulate the change in frequencies of a given allele (say A) in a diploid population of size N, starting from any arbitrary initial frequency p0.

This is easily done using the R package “learnPopGen”

You can also do this online visiting:
http://www.phytools.org/PopGen/ We will do the following sections:
3. drift-selection
4. founder-event
7. migration-selection-drift


# Exercise 1

Use the function drift.selection() to simulate the evolution of a population due to genetic drift.
Simulate the allele frequency changes at one single locus with 2 alleles, from a population of size 100, for 50 generations, with an initial allele frequency of 0.25.

Below you have the code for an example with an initial frequency of 0.1 (p0=0.1), from a population with an effective size of 100 (Ne=100), a single replicate (nrep=1 locus), for 100 generations (ngen=100).

NOTE: We are just simulating the effects of genetic drift on the allele frequencies of two alleles. There are no new mutations. There is no natural selection (both alleles have the same fitness). There is no migration. Thus, the fitness (w) of the 3 possible genotypes, namely DD (homozygote derived allele), AD (heterozygote), AA (homozygote ancestral allele), is given by 1, 1, 1, respectively (w=c(1,1,1)).

The output of the function drift.selection is a list with the allele trajectories.


```{r}
library("learnPopGen")
library("RColorBrewer")


 # Use variables to save the settings
  eSize <- 100 # effective size
  relFitness <- c(1,1,1) # relative fitness of DD, AD, AA genotypes, where A is the ancestral allele and D is the derived allele
  initFreq <- 0.25 # initial frequency of derived allele D at time zero 
  time <- 50 # number of generations
  
  # Simulate 1 SNP locus
  neutral_snp <- drift.selection(p0=initFreq, Ne=eSize, w=relFitness, nrep=1, ngen=time, show="p") 
  # Simulate 10 SNP loci using 10 colors generated with R Color Brewer
  mycol <- brewer.pal(n=10, name = "Paired")
  neutral_snp <- drift.selection(p0=initFreq, Ne=eSize, w=relFitness, nrep=10, ngen=time, show="p", colors = mycol[1:10]) 
  
  # Convert the list with output of the 10 simulations to a matrix
  # where each column is an independent simulation, and each row is a generation
  neutral_snp <- sapply(neutral_snp, function(x) {x}, simplify=TRUE)
  
  # Compute the mean frequency at each generation across simulations
  meanfreq_neutral <- rowMeans(neutral_snp)
  # Add a line with the mean frequency
  lines(0:time, meanfreq_neutral, lwd=3, lty=3, col="black")

```


# Exercise 2

Perform simulations with different Ne to answer the following questions:

* What is the effect of decreasing (Ne=10) or increasing (Ne=200) the effective size?
* How does Ne affect the probability of loosing the allele, i.e. reaching a frequency of 0.0?
* How does Ne affect the probability of fixation, i.e. reaching a frequency of 1.0?
* How does Ne affect the time it takes to reach fixation?


2N -> average time to fixation (binomial average)
time to fixation -> increases with the effective size

Initial freq = expected probability of fixation


```{r}
# Exercise 2; effect of decreasing or increasing Ne 

# Ne = 10

ne_10 = drift.selection(p=0.25, Ne=10, ngen=50, nrep=10)
ne_200 =drift.selection(p=0.25, Ne=200, ngen=50, nrep=10)

matrix = sapply(ne_10, function(x){x}, simplify = TRUE)
matrix = sapply(ne_10, function(x){x}, simplify = TRUE)

print(matrix)
try = which(matrix[,1]==0.0)
print(try[1])


lost = 0
fixed = 0
for (ne in seq(100,500, by=100)) {
  list = drift.selection(p=0.25, Ne=ne, ngen=200, nrep=10)
  matrix = sapply(list, function(x){x}, simplify = TRUE)
  if (any(matrix[,1]==0)){
    lost = lost +1 
  } 
  if (any(matrix[,1]==1)){
    fixed = fixed+1
  }
  meanfreq = rowMeans(matrix)
  lines(0:200, meanfreq, lwd=3, lty=3, col="black")
}
print(lost)
print(fixed)



```

  * What is the effect of decreasing (Ne=10) or increasing (Ne=200) the effective size?
    * As ne decreases, the probability of loosing or fix an allele increases and the more Ne increase, the longer it takes to reach a steady state. 
  
  * How does Ne affect the probability of loosing the allele, i.e. reaching a frequency of 0.0?
  * How does Ne affect the probability of fixation, i.e. reaching a frequency of 1.0?
    * Increasing Ne doesn't really increase the chance of fixation if initial allele frequency and fitness stay the same. In general, the higher the effective size Ne, the lower are the effect of drift and the longer it will take to have a loss or a fixation.
  
  How does Ne affect the time it takes to reach fixation?
  * The bigger the population, the longer it will take to have an allele fixation. 
  
  
  










# Exercise 3

Perform simulations with different Ne to answer the following questions:

* What is the effect of changing the initial frequency? Simulate the trajectory of a new mutation with initial frequency of 1/(2*Ne).

```{r}
# Exercise 3 : chaning NE 
lost = 0
fixed = 0
for (ne in seq(10,500, by=20)) {
  f = 1/(2*ne)
  list = drift.selection(p=f, Ne=ne, ngen=200, nrep=10)
  matrix = sapply(list, function(x){x}, simplify = TRUE)
  if (any(matrix[,1]==0)){
    lost = lost +1 
    print(paste("lost at ne of", ne))
    
    lost_at = which(matrix[,1]==0)
    print(paste("lost at generation", lost_at[1]))
    
  } 
  if (any(matrix[,1]==1)){
    fixed = fixed+1
    print(paste("lost at ne of", ne))
    
    fixed_at = which(matrix[,1]==1)
    print(paste("fixed at generation", fixed_at[1]))
    
  }
  meanfreq = rowMeans(matrix)
  lines(0:200, meanfreq, lwd=3, lty=3, col="black")
}
print(lost)
print(fixed)

```
Such a small Ne makes it really likely to loose the allele 


* How does the initial frequency affect the probability of fixation? -> it's the same probability! 

```{r}
# Exercise 3: changing po
lost = 0
fixed = 0
ne = 200
for (f in seq(0,1, by=0.1)) {
  list = drift.selection(p=f, Ne=ne, ngen=200, nrep=10)
  matrix = sapply(list, function(x){x}, simplify = TRUE)
  if (any(matrix[,1]==0)){
    lost = lost +1 
    print(paste("lost at p0 of ", toString(f)))
    
    lost_at = which(matrix[,1]==0)
    print(paste("lost at generation", lost_at[1]))
    
  } 
  if (any(matrix[,1]==1)){
    fixed = fixed+1
    print(paste("fixed at p0 of", toString(f)))
    
    fixed_at = which(matrix[,1]==1)
    print(paste("fixed at generation ", fixed_at[1]))
    
  }
  meanfreq = rowMeans(matrix)
  lines(0:200, meanfreq, lwd=3, lty=3, col="black")
}
print(lost)
print(fixed)

```


* How does the initial frequency affect the time to fixation?

The closer to 100% it is, the more chance it has to be fixed. 


TO DISCUSS:

* How many times was the allele lost? Did this depend on Ne and p0?

The lower the initial frequency the bigger the allele will be lost. 

* How many times was the allele fixed in the population? After how many generations does fixation occur? Did this depend on Ne and p0?

Time of fixation -> linear relationship with time and Ne ! The initial frequency does not change anything (seems complicated to prove theoretically but with simulation we see it's true)

* How did the mean frequency across simulations changed through time?

The mean frequency remain constant accross the generation time. 


* What is the most likely fate of a new mutation entering a population?

Since the initial freq is low, it will most likely be lost.


* How do the genotypic frequencies change? Are there deviations from Hardy-Weinberg proportions due to genetic drift?

We see that the allele freq are changing -> thus proportion of homo/heterozygous are also changing. No deviation from hardy weinberg. All individuals are reproducing at random ! After one generartion we are in HWE. 

* What data would we need to test predictions from genetic drift?

We need a time series data (like Esra's flies !!! -> 7 years of evolution experiment on Drosophila) 















# POSITIVE SELECTION

In the following exercises, we will perform simulations with a beneficial deived allele with a selection coefficient of s. That means that individuals that are homozygote for the D allele (i.e., DD) will have a fitness of w=1+s.

## Exercise 4
  Perform simulations to answer the following questions:
  - What is the effect of directional selection on allele frequency trajectories?
  - Is the fate of a beneficial allele the same in a small and large population?





```{r}
# Exercise 4 

size = 10
h = 0.5
s = 0.2
relFitness = c(1+s, 1+(h*s), 1)
initFreq= 0.25
time = 100

sel_snp = drift.selection(p0=initFreq, Ne=size, w=relFitness, nrep = 1, ngen=time, show="p")
sel_snp = sapply(sel_snp, function(x) {x}, simplify = TRUE)
meanfreq_sel = rowMeans(sel_snp)
lines(0:time, meanfreq_sel, lwd=3, lty=3, col="black")




```


# Exercise 5

Perform simulations to answer the following questions:

- Is the trajectory of a new mutation with selective coefficient s similar in a large and small population?



Compare the trajectory of a beneficially selected allele at a biallelic SNP in a population with Ne=10 with a population with Ne=200, where the initial frequency is p0=1/(2Ne). This mimicks the dynamics of a new beneficial mutation entering the population. Consider a co-dominant (h=0.5) beneficial mutation with selective coefficient s=0.05.

Consider different values for the selective coefficient, ranging from s=0 (neutral) to s=0.50 (strongly beneficial).




1.050 -> this homozygous have 5% more offspring
1.025 -> heterozygous have half of the homozygous fitness advantage 
  if 1.050 -> codominant. 
1.000 -> no advantage in homozygous without the mutation




```{r}
# Compare trajectories with p = 1/2ne 
# h = 0.5 and s = 0.05
# of populations 10 and 200 
size = 10
h = 0.5
s = 0.05
relFitness = c(1+s, 1+(h*s), 1)
initFreq = 1/(2*size) 
time=100

sel_snp = drift.selection(p0=initFreq, Ne=size, w=relFitness, nrep = 100, ngen=time, show="p")
sel_snp = sapply(sel_snp, function(x) {x}, simplify = TRUE)
meanfreq_sel = rowMeans(sel_snp)
lines(0:time, meanfreq_sel, lwd=3, lty=3, col="black")

size = 200
h = 0.5
s = 0.5
relFitness = c(1+s, 1+(h*s), 1)
initFreq = 1/(2*size) 
time=100

sel_snp = drift.selection(p0=initFreq, Ne=size, w=relFitness, nrep = 100, ngen=time, show="p")
sel_snp = sapply(sel_snp, function(x) {x}, simplify = TRUE)
meanfreq_sel = rowMeans(sel_snp)
lines(0:time, meanfreq_sel, lwd=3, lty=3, col="black")


# Consider different values of s 
ne = 10
lost = 0
fixed = 0
for (s in seq(0,0.5,by=0.1)){
  p = 1/(2*ne)
  relFitness = c(1+s, 1+(h*s), 1)
  sel_snp = drift.selection(p0=p, Ne=size, w=relFitness, nrep = 10, ngen=time, show="p")
  sel_snp = sapply(sel_snp, function(x) {x}, simplify = TRUE)
  if (any(sel_snp[,1]==0)){
    lost = lost +1 
    print("lost at s of")
    print(s)
    print("lost at generation")
    lost_at = which(sel_snp[,1]==0)
    print(lost_at[1])
  } 
  if (any(sel_snp[,1]==1)){
    fixed = fixed+1
    print("fixed at s of")
    print(s)
    print("fixed at generation")
    fixed_at = which(sel_snp[,1]==1)
    print(fixed_at[1])
  }
  meanfreq = rowMeans(sel_snp)
  lines(0:100, meanfreq, lwd=3, lty=3, col="black")
}
print(lost)
print(fixed)


# Consider different values of ne and s 
lost = 0
fixed = 0
for (s in seq(0,0.5,by=0.1)){
  relFitness = c(1+s, 1+(h*s), 1)
  for (ne in seq(50,500,by=50)){
    p = 1/(2*ne)
    sel_snp = drift.selection(p0=p, Ne=ne, w=relFitness, nrep = 10, ngen=time, show="p")
    sel_snp = sapply(sel_snp, function(x) {x}, simplify = TRUE)
    if (any(sel_snp[,1]==0)){
      lost = lost +1 
      print("lost at ne of")
      print(ne)
      print("at s of")
      print(s)
      print("lost at generation")
      lost_at = which(sel_snp[,1]==0)
      print(lost_at[1])
    } 
    if (any(sel_snp[,1]==1)){
      fixed = fixed+1
      print("lost at ne of")
      print(ne)
      print("fixed at s of")
      print(s)
      print("fixed at generation")
      fixed_at = which(sel_snp[,1]==1)
      print(fixed_at[1])
    }
    meanfreq = rowMeans(sel_snp)
    lines(0:100, meanfreq, lwd=3, lty=3, col="black")
  }
}
  
print(lost)
print(fixed)

```


Selection only works if selective coefficient is larger than 1/2N 
Some mutation can have different effect




## TO DISCUSS:

* How often is the beneficial allele lost? How is the probability of loss affected by the effective size of populations?




* What is the impact of the effective population size on the probability of fixation of a new beneficial allele?


* What is the impact of drift and selection on the mean frequency across simulations? What is the impact of drift and selection on the variance across simulations?


* How to distinguish drift from selection? - What is the impact of demography on the efficiency of selection?


# DEMOGRAPHIC EVENTS

The next exercise simulates the effect of a bottleneck in the allele frequency trajectories of a neutral mutation.

# Exercise 6
What is the effect of a bottleneck or founder event?
Use the function founder.event()


















